esp-bootloader-esp-idf/src/partitions.rs

```diff
@@ -0,0 +1,757 @@
+    /// Used for Non-Volatile Storage (NVS).
+    Nvs,
+    /// Used for storing core dumps while using a custom partition table
+    Coredump,
+    /// NvsKeys is used for the NVS key partition. (NVS).
+    NvsKeys,
+    /// Used for emulating eFuse bits using Virtual eFuses.
+    EfuseEm,
+    /// Implicitly used for data partitions with unspecified (empty) subtype,
+    /// but it is possible to explicitly mark them as undefined as well.
+    Undefined,
+    /// FAT Filesystem Support.
+    Fat      = 0x81,
+    /// SPIFFS Filesystem.
+    Spiffs   = 0x82,
+    ///  LittleFS filesystem.
+    LittleFs = 0x83,
+}
+
+impl TryFrom<u8> for DataPartitionSubType {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        DataPartitionSubType::from_repr(value).ok_or(Error::Invalid)
+    }
+}
+
+/// Sub-type of the bootloader partition type.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::FromRepr)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[repr(u8)]
+pub enum BootloaderPartitionSubType {
+    /// It is the so-called 2nd stage bootloader.
+    Primary = 0,
+    /// It is a temporary bootloader partition used by the bootloader OTA update
+    /// functionality for downloading a new image.
+    Ota     = 1,
+}
+
+impl TryFrom<u8> for BootloaderPartitionSubType {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        BootloaderPartitionSubType::from_repr(value).ok_or(Error::Invalid)
+    }
+}
+
+/// Sub-type of the partition table type.
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::FromRepr)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+#[repr(u8)]
+pub enum PartitionTablePartitionSubType {
+    /// It is the primary partition table.
+    Primary = 0,
+    /// It is a temporary partition table partition used by the partition table
+    /// OTA update functionality for downloading a new image.
+    Ota     = 1,
+}
+
+impl TryFrom<u8> for PartitionTablePartitionSubType {
+    type Error = Error;
+
+    fn try_from(value: u8) -> Result<Self, Self::Error> {
+        PartitionTablePartitionSubType::from_repr(value).ok_or(Error::Invalid)
+    }
+}
+
+/// Read the partition table.
+///
+/// Pass an implementation of [embedded_storage::Storage] which can read from
+/// the whole flash and provide storage to read the partition table into.
+pub fn read_partition_table<'a>(
+    flash: &mut impl embedded_storage::Storage,
+    storage: &'a mut [u8],
+) -> Result<PartitionTable<'a>, Error> {
+    flash
+        .read(PARTITION_TABLE_OFFSET, storage)
+        .map_err(|_e| Error::StorageError)?;
+
+    PartitionTable::new(storage)
+}
+
+/// A flash region is a "view" into the partition.
+///
+/// It allows to read and write to the partition without the need to account for
+/// the partition offset.
+#[derive(Debug)]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
+pub struct FlashRegion<'a, F> {
+    raw: &'a PartitionEntry<'a>,
```

Comment:
imo this can be stored by value.
```suggestion
    raw: PartitionEntry<'a>,
```

Reply:
I'd prefer to keep it zero-copy